content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+
+The source code and the details about the setup can be found in [github repository](https://github.com/narqo/postmortem-debug-go). To follow along, you will need a VM running Linux. I will use Vagrant with hostmanager plugin. Refer to Vagrantfile in the root of the repository.
+
+To debug the problem, we need to bump into the problem first. Let’s start the VM, build our HTTP service, run it and see what will happen.
+
+```
+= vagrant ssh server-test-1
+Welcome to Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-33-generic x86_64)
+
+:~$ cd /vagrant/example/server
+:/vagrant/example/server$ go build -o server ./
+:/vagrant/example/server$ ./server --addr=:10080
+
+server listening addr=:10080
+```
+
+Using [wrk][] HTTP benchmarking tool start adding some load. I run this demo on my MacBook with four cores. Running wrk with four threads and 1000 connections is enough to simulate the failure we want to debug. Run the following command in a new terminal panel:
+
+```
+= wrk -d1m -t4 -c1000 'http://server-test-1:10080'
+Running 1m test @ http://server-test-1:10080
+  4 threads and 1000 connections
+  ···
+```
+
+After a brief period, the server gets stacked. Even after wrk finished the run, the service is unable to process an incoming request:
+
+```
+= curl --max-time 5 'http://server-test-1:10080/'
+curl: (28) Operation timed out after 5001 milliseconds with 0 bytes received
+```
+
+Indeed, we have a problem! Let have a look.
+
+---
+
+In a similar situation with our production service, after a brief period, the total number of spawned goroutines for incoming requests has risen so much, the server became unresponsive. Even requests to pprof debug handlers were *s-u-u-u-per slow*, making it look like the server was completely "dead". Similarly, our attempts to kill the process with `SIGQUIT` to [get the stack dump of running goroutines][1] didn't seem to work.
+
+---
+
+## GDB and Coredump
+
+We can start with trying to inspect the running service with GDB (GNU Debugger).
+
+*Running a debugger in the production environment will likely require additional privileges. If in doubt, be wise and always consult with your operations team first.*
+
+Connect to another SSH session on the VM, find server process's id, and attach to the process with the debugger:
+
+```
+= vagrant ssh server-test-1
+Welcome to Ubuntu 18.04.1 LTS (GNU/Linux 4.15.0-33-generic x86_64)
+
+:~$ ps -ef | grep server
+vagrant   1628  1557  0 14:45 pts/0	00:00:01 ./server --addr=:10080
+
+:~$ sudo gdb --pid=1628 /vagrant/example/server/server
```

Comment:
I would add a `cd /vagrant` so that the core file is in the right directory for later and can be analysed from outside the VM.